File: src/types/size.ts
import { Point } from "./point";
import { u32 } from "./u32";
class Size {
  constructor(
    public width: u32,
    public height: u32,
  ) {}
  clone(): Size {
    return new Size(this.width, this.height);
  }
  public toCode(): string {
    return `Size(width=${this.width}, height=${this.height})`;
  }
  static fromPoints(start: Point, end: Point): Size {
    const width = Math.abs(end.x - start.x);
    const height = Math.abs(end.y - start.y);
    return new Size(width, height);
  }
}
export { Size };

File: src-tauri/src/types/size.rs
// done
use super::Point;
use anyhow::{Result, anyhow};
use serde::{Deserialize, Serialize};
#[derive(Debug, Serialize, Deserialize, Clone, Copy, PartialEq)]
pub struct Size {
    pub width: u32,
    pub height: u32,
}
impl Size {
    pub fn new(width: impl Into<u32>, height: impl Into<u32>) -> Self {
        Self {
            width: width.into(),
            height: height.into(),
        }
    }
    pub fn new_from_start_end_points(start_point: Point, end_point: Point) -> Result<Self> {
        let width = end_point.x - start_point.x;
        let height = end_point.y - start_point.y;
        if width <= 0 {
            return Err(anyhow!(t!("Width must be greater than zero.")));
        }
        if height <= 0 {
            return Err(anyhow!(t!("Height must be greater than zero.")));
        }
        let (width, height) = (width as u32, height as u32);
        Ok(Self { width, height })
    }
}
